# LOADING AND CLEANING THE DATA


In [4]:
import sys
sys.path.append( './modules') #module folder contains the helper file(a collection of created functions). see below
import pandas as pd
import joblib
import helper #helper module contains functions that were created to process the data
from helper import upload_file_csv, case_check, summary_stat, missing_vals, drop_missing,  check_dup, drop_feature, merger, drop_dup

In [ ]:
importlib.reload(hlp)

In [1]:
#print(upload_file_csv.__doc__)

In [5]:
#get dataset paths
train_features_link = "DSDJ-data-SalaryPredictions/data/train_features.csv"
train_target_link = "DSDJ-data-SalaryPredictions/data/train_salaries.csv"

In [6]:
train_features = upload_file_csv(train_features_link)
train_features

jobId companyId         jobType       degree      major  \
0       JOB1362684407687    COMP37             CFO      MASTERS       MATH   
1       JOB1362684407688    COMP19             CEO  HIGH_SCHOOL       NONE   
2       JOB1362684407689    COMP52  VICE_PRESIDENT     DOCTORAL    PHYSICS   
3       JOB1362684407690    COMP38         MANAGER     DOCTORAL  CHEMISTRY   
4       JOB1362684407691     COMP7  VICE_PRESIDENT    BACHELORS    PHYSICS   
...                  ...       ...             ...          ...        ...   
999995  JOB1362685407682    COMP56  VICE_PRESIDENT    BACHELORS  CHEMISTRY   
999996  JOB1362685407683    COMP24             CTO  HIGH_SCHOOL       NONE   
999997  JOB1362685407684    COMP23          JUNIOR  HIGH_SCHOOL       NONE   
999998  JOB1362685407685     COMP3             CFO      MASTERS       NONE   
999999  JOB1362685407686    COMP59          JUNIOR    BACHELORS       NONE   

         industry  yearsExperience  milesFromMetropolis  
0          HEALTH               10                   83  
1             WEB                3                   73  
2          HEALTH               10                   38  
3            AUTO                8                   17  
4         FINANCE                8                   16  
...           ...              ...                  ...  
999995     HEALTH               19                   94  
999996    FINANCE               12                   35  
999997  EDUCATION               16                   81  
999998     HEALTH                6                    5  
999999  EDUCATION               20                   11  

[1000000 rows x 8 columns]

In [49]:
train_target = upload_file_csv(train_target_link)
train_target

jobId  salary
0       JOB1362684407687     130
1       JOB1362684407688     101
2       JOB1362684407689     137
3       JOB1362684407690     142
4       JOB1362684407691     163
...                  ...     ...
999995  JOB1362685407682      88
999996  JOB1362685407683     160
999997  JOB1362685407684      64
999998  JOB1362685407685     149
999999  JOB1362685407686      88

[1000000 rows x 2 columns]

-----

In [50]:
summary_stat(train_features) #summary stats of feature data set

Data has 1000000 rows by 8 columns

Data type
jobId                  object
companyId              object
jobType                object
degree                 object
major                  object
industry               object
yearsExperience         int64
milesFromMetropolis     int64
dtype: object


jobId companyId  jobType       degree    major industry  \
count            1000000   1000000  1000000      1000000  1000000  1000000   
unique           1000000        63        8            5        9        7   
top     JOB1362685353242    COMP39   SENIOR  HIGH_SCHOOL     NONE      WEB   
freq                   1     16193   125886       236976   532355   143206   
mean                 NaN       NaN      NaN          NaN      NaN      NaN   
std                  NaN       NaN      NaN          NaN      NaN      NaN   
min                  NaN       NaN      NaN          NaN      NaN      NaN   
25%                  NaN       NaN      NaN          NaN      NaN      NaN   
50%                  NaN       NaN      NaN          NaN      NaN      NaN   
75%                  NaN       NaN      NaN          NaN      NaN      NaN   
max                  NaN       NaN      NaN          NaN      NaN      NaN   

        yearsExperience  milesFromMetropolis  
count    1000000.000000       1000000.000000  
unique              NaN                  NaN  
top                 NaN                  NaN  
freq                NaN                  NaN  
mean          11.992386            49.529260  
std            7.212391            28.877733  
min            0.000000             0.000000  
25%            6.000000            25.000000  
50%           12.000000            50.000000  
75%           18.000000            75.000000  
max           24.000000            99.000000

Features set looks as expected . 

YearsExperience and milesFromMetropolis are integers and other features are strings

jobid has as many unique values as number of rows so it will  probably have no predictive use

In [51]:
summary_stat(train_target) #summary stats of feature data set

Data has 1000000 rows by 2 columns

Data type
jobId     object
salary     int64
dtype: object


jobId          salary
count            1000000  1000000.000000
unique           1000000             NaN
top     JOB1362685353242             NaN
freq                   1             NaN
mean                 NaN      116.061818
std                  NaN       38.717936
min                  NaN        0.000000
25%                  NaN       88.000000
50%                  NaN      114.000000
75%                  NaN      141.000000
max                  NaN      301.000000

Again jobid in target data set has as many unique values as number of rows and will probably have no predictive use

In [52]:
case_check(train_features) # function checks if each feature values (object types) all have uniform cases. It makes them all uppercase if not

Checking if case in feature values are uniform
jobId ---> All upppercase
companyId ---> All upppercase
jobType ---> All upppercase
degree ---> All upppercase
major ---> All upppercase
industry ---> All upppercase



In [53]:
missing_vals(train_features) #checks for missing values in features dataframe

Percent missing of total
jobId                  0.0
companyId              0.0
jobType                0.0
degree                 0.0
major                  0.0
industry               0.0
yearsExperience        0.0
milesFromMetropolis    0.0
dtype: float64


In [54]:
missing_vals(train_target)  #checks for missing values in target dataframe

Percent missing of total
jobId     0.0
salary    0.0
dtype: float64


In [55]:
check_dup(train_features)

0 duplicates found


In [56]:
check_dup(train_target)

0 duplicates found


No duplicated in feature and target data set as expected since jobid entries are unique in both data sets

In [57]:
#checking duplicates after droping jobId in features train set since jobid has no predictive use
check_dup(drop_feature(train_features, 'jobId' ))

8001 duplicates found


8001 duplicates were found after droping jobId. We might need to explore and decide what to do with this during EDA

----------------

In [58]:
#merging feature and target train sets
merged_train_data = merger(train_features, train_target, 'jobId', how= 'left')
merged_train_data

jobId companyId         jobType       degree      major  \
0       JOB1362684407687    COMP37             CFO      MASTERS       MATH   
1       JOB1362684407688    COMP19             CEO  HIGH_SCHOOL       NONE   
2       JOB1362684407689    COMP52  VICE_PRESIDENT     DOCTORAL    PHYSICS   
3       JOB1362684407690    COMP38         MANAGER     DOCTORAL  CHEMISTRY   
4       JOB1362684407691     COMP7  VICE_PRESIDENT    BACHELORS    PHYSICS   
...                  ...       ...             ...          ...        ...   
999995  JOB1362685407682    COMP56  VICE_PRESIDENT    BACHELORS  CHEMISTRY   
999996  JOB1362685407683    COMP24             CTO  HIGH_SCHOOL       NONE   
999997  JOB1362685407684    COMP23          JUNIOR  HIGH_SCHOOL       NONE   
999998  JOB1362685407685     COMP3             CFO      MASTERS       NONE   
999999  JOB1362685407686    COMP59          JUNIOR    BACHELORS       NONE   

         industry  yearsExperience  milesFromMetropolis  salary  
0          HEALTH               10                   83     130  
1             WEB                3                   73     101  
2          HEALTH               10                   38     137  
3            AUTO                8                   17     142  
4         FINANCE                8                   16     163  
...           ...              ...                  ...     ...  
999995     HEALTH               19                   94      88  
999996    FINANCE               12                   35     160  
999997  EDUCATION               16                   81      64  
999998     HEALTH                6                    5     149  
999999  EDUCATION               20                   11      88  

[1000000 rows x 9 columns]

In [59]:
# checking for duplicates on merged data after dropping jobId
check_dup(drop_feature(merged_train_data, 'jobId' ))

186 duplicates found


After droping jobId, only 186 duplicates were found on the merged data compared with 8001 duplicates found on features train set after droping jobId.

This means excluding jobId from the features, the other features have numerous combinations that appear multiple times but with different salary values.

Possibly there are missing information or jobId carries information that is not apparent

We will explore this in EDA

-----

First line of code below removes the rows in the train_features data that corresponds to the duplicate rows in the merged data set

The second line of code gets us the data that we need to explore. It contains feature combinations( after excluding jobId) that are duplicated but with different salary value 

The duplicates show below includes the 'original' and we are going to explore this in EDA


In [60]:
train_features_clnd = train_features.drop(train_features[drop_feature(merged_train_data , 'jobId' ).duplicated(keep = 'first')== True].index)
dupl_merged_data = clean_merged_data [drop_feature(train_features_clnd , 'jobId' ).duplicated(keep = False)== True]

<ipython-input-60-60d73c48ef69>:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  dupl_merged_data = clean_merged_data [drop_feature(train_features_clnd , 'jobId' ).duplicated(keep = False)== True]


In [61]:
dupl_merged_data  #rows with duplicate feature combination but different salary value

companyId  jobType       degree      major industry  yearsExperience  \
83        COMP17  MANAGER         NONE       NONE  SERVICE                8   
122       COMP46  JANITOR         NONE       NONE     AUTO                1   
216       COMP36  JANITOR         NONE       NONE      OIL               16   
294       COMP37      CEO     DOCTORAL  CHEMISTRY     AUTO               11   
302        COMP1  JANITOR         NONE       NONE   HEALTH                8   
...          ...      ...          ...        ...      ...              ...   
999690     COMP0      CEO  HIGH_SCHOOL       NONE  FINANCE               22   
999791    COMP43      CEO  HIGH_SCHOOL       NONE  SERVICE                3   
999825    COMP40   SENIOR         NONE       NONE      OIL                6   
999830     COMP6      CEO         NONE       NONE     AUTO               18   
999923    COMP40   SENIOR         NONE       NONE      OIL                6   

        milesFromMetropolis  salary  
83                       15     111  
122                      21      55  
216                      87      72  
294                      60     135  
302                      38      73  
...                     ...     ...  
999690                   32     183  
999791                   25     101  
999825                   24      86  
999830                   15     150  
999923                   24     112  

[15548 rows x 8 columns]

In [62]:
merged_train_data = drop_missing(merged_train_data, 'salary') #drops rows with missing values including empty strings and  salary values less than or equal to zero

Percent missing of total
jobId                  0.0
companyId              0.0
jobType                0.0
degree                 0.0
major                  0.0
industry               0.0
yearsExperience        0.0
milesFromMetropolis    0.0
salary                 0.0
dtype: float64
Output After dropping missing values
There are 999995 entries with non-zero salary values


In [63]:
clean_merged_data= drop_dup(drop_feature(merged_train_data, 'jobId' )) #this code removes rows that are duplicated in the merged data

186 duplicates found and removed


In [64]:
joblib.dump(clean_merged_data, 'merged_part_cleaned_trainset') #saving cleaned merged feature and target data set

['merged_part_cleaned_trainset']

In [65]:
joblib.dump(dupl_merged_data, 'Duplicates_in_trainset') #saving duplicate feature combination in merged data

['Duplicates_in_trainset']